KK_ButtPlugin/ButtplugController.cs

```diff
@@ -102,22 +157,37 @@ IEnumerator RunVibrate()
         {
             while (!flags.isHSceneEnd)
             {
-                if (flags.nowAnimStateName.Equals("OLoop"))
-                {
-                    DoVibrate(1.0f);
-                }
-                else if (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))
+                if (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))
                 {
                     // stops vibration when not being lewd
                     DoVibrate(0.0f);
+                    yield return new WaitForSecondsRealtime(1.0f / (float)ButtPlugin.VibrationFrequency.Value);
+                    continue;
                 }
-                else
+
+                var animator = flags.lstHeroine[0].chaCtrl.animBody;
+
+                var speed = flags.speedCalc;
+                var strength = 1.0f;
+                var minVibration = 0.2f;
+
+                // service mode goes into OLoop once male excitement exceeds its threshold
+                if (IsOrgasm)
                 {
-                    // vibrate based on the intensity of the player
-                    // minimum vibration above 0 exists so you always feel something along with the animation
-                    DoVibrate(Mathf.Lerp(0.2f, 1.0f, flags.speedCalc));
+                    speed = 1.0f;
+                    minVibration = 0.6f;
                 }
-                yield return new WaitForSeconds(.01f);
+
+                if (ButtPlugin.SyncVibrationWithAnimation.Value)
+                {
+                    // Simple sin based intensity amplification based on normalized position in looping animation
+                    var info = animator.GetCurrentAnimatorStateInfo(0);
+                    var depth = (info.m_NormalizedTime - GetPhase()) % 1;

```

Comment:
Edge case: in the beginning of the animation this can be negative because of modulo weirdness, and then strength will be negative too.

Reply:
Thankfully lerp later will compensate for that, as any values outside the range gets clamped to the ends, so a negative value will just end up mapping to minVibration, similar to a zero.

Reply:
Absolutely right, my bad